macrosynergy/pnl/historic_portfolio_volatility.py

```diff
@@ -210,69 +329,62 @@ def _calculate_portfolio_volatility(
+    )
+
     for td in rebal_dates:
-        window_df = pivot_returns.loc[pivot_returns.index <= td]
-        dict_vcv: Dict[str, pd.DataFrame] = {
-            freq: estimate_variance_covariance(
-                piv_ret=_downsample_returns(window_df, freq=freq).iloc[
-                    -get_max_lookback(lb=lb, nt=nan_tolerance) :
-                ],
-                lback_periods=lb,
-                remove_zeros=remove_zeros,
-                weights_func=weights_func,
-                half_life=hl,
+        avails = first_starts[first_starts <= td].index.tolist()
+        if len(avails) == 0:
+            logger.warning(
+                f"No data available for {td} with lookback period of {max(lback_periods)} days."
             )
-            for freq, lb, hl in zip(est_freqs, lback_periods, half_life)
-        }
-
-        vcv_df: pd.DataFrame = sum(
-            [
-                est_weights[ix] * ANNUALIZATION_FACTORS[freq] * dict_vcv[freq]
-                for ix, freq in enumerate(est_freqs)
-            ]
-        )
-        # reshape and drop duplicates in vcv
-
-        list_vcv.append(
-            vcv_df.rename_axis("fid1", axis=0)
-            .rename_axis("fid2", axis=1)
-            .stack()
-            .to_frame("value")
-            .reset_index()
-            .assign(real_date=td)
+            continue
+        vcv_df = _calculate_multi_frequency_vcv_for_period(
+            pivot_returns=pivot_returns[avails],
+            pivot_signals=pivot_signals[avails],
+            rebal_date=td,
+            est_freqs=est_freqs,
+            est_weights=est_weights,
+            weights_func=weights_func,
+            lback_periods=lback_periods,
+            half_life=half_life,
+            nan_tolerance=nan_tolerance,
+            remove_zeros=remove_zeros,
         )
 
-        # TODO adjust for signal zero or N/A
-        s = signals.loc[td, :]
-        mask: pd.Series = s.isnull() | (s.abs() < 1e-8)
-        vcv_df.loc[s.index[mask], :] = 0
-        vcv_df.loc[:, s.index[mask]] = 0
-        s[mask] = 0
-        assert (
-            vcv_df.isnull().sum().sum() == 0
-        ), f"N/A values in variance-covariance matrix!\n{vcv_df}"
-
-        pvol: float = np.sqrt(s.T.dot(vcv_df).dot(s))
-        list_pvol.append((td, pvol))
+        list_vcv.append(stack_covariances(vcv_df=vcv_df, real_date=td))
+        vol_tuple = _calc_vol_tuple(
+            vcv_df=vcv_df,
+            signals=signals,
+            date=td,
+            available_cids=avails,
+        )
+        list_pvol.append(vol_tuple)
 
     pvol = pd.DataFrame(
         list_pvol,
         columns=["real_date", portfolio_return_name],
     ).set_index("real_date")
 
-    vcv_df = pd.concat(list_vcv, axis=0)  # add to cls.vcv
+    vcv_df_long = pd.concat(list_vcv, axis=0)  # add to cls.vcv
 
-    vcv_df["helper"] = vcv_df[["fid1", "fid2", "real_date"]].apply(
+    vcv_df_long["helper"] = vcv_df_long[["fid1", "fid2", "real_date"]].apply(
```

Comment:
This is on me, but lets find a better naming convention than `fid1` and `fid2`, maybe `fid_rows` and `fid_cols`?